src/v/rpc/service.h

```diff
@@ -81,6 +82,14 @@ struct service::execution_helper {
       method_info method,
       Func&& f) {
         return ctx.permanent_memory_reservation(ctx.get_header().payload_size)
+          .handle_exception([&ctx](const std::exception_ptr& e) {
+              // It's possible to stop all waiters on a semaphore externally
+              // with the semaphore's `broken` method. In which case
+              // `permanent_memory_reservation` will return an exception.
+              // We intercept it here to avoid a broken promise.
+              ctx.body_parse_exception(e);
+              std::rethrow_exception(e);
```

Comment:
nit: it is more idiomatic here outside a co-routine context to `return ss::make_exception_future<>(e)`. the reason is that seastar will end up having to catch this again and  repackage it into an exceptional future.

Reply:
Good point, switching to it now.